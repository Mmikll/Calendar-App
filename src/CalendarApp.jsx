import { BrowserRouter } from 'react-router-dom';
import { AppRouter } from './router';
import { Provider } from 'react-redux';
import { store } from './store';

export const CalendarApp = () => {
  return (
    <div>
      <Provider store={store}>
        <BrowserRouter>
          <AppRouter/>
        </BrowserRouter>
      </Provider>

    </div>
  )
}


