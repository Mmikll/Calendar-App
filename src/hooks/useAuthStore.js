import { useDispatch, useSelector } from "react-redux"
import calendarApi from '../api/calendarApi'
import { clearErrorMessage, onChecking, onLogin, onLogout } from "../store/auth/authSlice"

export const useAuthStore = () => {

    const { status, user, errorMessage } = useSelector( state => state.auth )
    const dispatch = useDispatch()

    const startLogin = async({ email, password }) => {
        dispatch( onChecking() ) 
        try {
            const { data } = await calendarApi.post('/auth/', { email, password })
            console.log({ data })

            // Guardar el token en localStorage
            localStorage.setItem('data', data.token)
            localStorage.setItem('data-init-date', new Date().getTime())

            dispatch(onLogin({
                name: data.name,
                uid: data.uid,
            }))
        } catch (err) {
            dispatch( onLogout("Not valid Credentials"))
            setTimeout(() => {
                dispatch(clearErrorMessage())
            },3000)
        }
    }

    const startRegister = async({ email, password, name }) => {
        dispatch( onChecking() ) 
        try {
            const { data } = await calendarApi.post('/auth/new', { email, password, name })
            console.log({ data })

            localStorage.setItem('data', data.token)
            localStorage.setItem('data-init-date', new Date().getTime())

            dispatch(onLogin({
                name: data.name,
                uid: data.uid,
            }))
        } catch (err) {
            dispatch( onLogout(err.response.data?.msg || 'There was an error'))
            setTimeout(() => {
                dispatch(clearErrorMessage())
            },3000)
        }
    }

    const checkAuthToken = async() => {

        const token = localStorage.getItem('token')
        if(!token) return dispatch( onLogout())
        
        try {
            const { data } = await calendarApi.post('/auth/renew')

            localStorage.setItem('data', data.token)
            localStorage.setItem('data-init-date', new Date().getTime())

            dispatch(onLogin({
                name: data.name,
                uid: data.uid,
            }))

        } catch (error) {
            localStorage.clear()
            dispatch( onLogout())
        }

    }

    const startLogout = () => {
        localStorage.clear()
        dispatch(onLogout())
    }

    return {
        // Propiedades
        status,
        user,
        errorMessage,

        // MÃ©todos
        startLogin,
        startRegister,
        checkAuthToken,
        startLogout,
    }
}
