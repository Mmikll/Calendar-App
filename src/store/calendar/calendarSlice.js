import { createSlice } from '@reduxjs/toolkit';

// const tempEvent = {
//   id: "12342323",
//     title: 'Boss Birthday',
//     notes: 'Must buy a cake',
//     start: new Date(2024, 9, 20, 10, 30),
//     end: new Date(2024, 9, 20, 12, 0),
//     bgColor: '#48e',
//     user: {
//       id: '123',
//       name: 'John Doe',
//     },
//   }

export const calendarSlice = createSlice({
    name: 'calendar',
    initialState: { 
        isLoadingEvents: true,
        events: [
            //tempEvent
        ],
        activeEvent: null,
    },
    reducers: { 
      onSetActiveEvent: (state, {payload}) =>{
        state.activeEvent = payload
      },
      onAddNewEvent: (state, {payload}) => {
        state.events.push(payload)
        state.activeEvent = null
      },
      onUploadNewEvent: (state, {payload}) => {
        state.events = state.events.map( event =>{
          if ( event.id === payload.id) {
            return payload
          }
          return event
        })
      },
      onDeleteEvent: (state) => {
        if ( state.activeEvent ){
          state.events = state.events.filter( event => event.id !== state.activeEvent.id)
          state.activeEvent = null
        }
      },
      onLoadEvents: (state, {payload = []}) => {
        state.isLoadingEvents = false

        payload.forEach( event => {
          const exist = state.events.some( (dbEvent) => dbEvent.id === event.id)
          if (!exist) {
            state.events.push(event)
          }
        })
      },
      onLogoutCalendar: (state) => {
        state.isLoadingEvents = true,
        state.events = [],
        state.activeEvent = null
      }

    },
});

export const {
  onSetActiveEvent,
  onAddNewEvent,
  onUploadNewEvent,
  onDeleteEvent,
  onLoadEvents,
  onLogoutCalendar } = calendarSlice.actions;