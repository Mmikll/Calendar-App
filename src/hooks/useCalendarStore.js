import { useDispatch, useSelector } from "react-redux"
import { onAddNewEvent, onDeleteEvent, onLoadEvents, onSetActiveEvent, onUploadNewEvent } from "../store/calendar/calendarSlice"
import calendarApi from "../api/calendarApi"
import { convertEventDate } from "../helpers"
import Swal from "sweetalert2"


export const useCalendarStore = () => {
    const dispatch = useDispatch()
    const {events, activeEvent} = useSelector( state => state.calendar)
    const {user} = useSelector( state => state.auth)

    const setActiveEvent = ( calendarEvent) =>{
        dispatch(onSetActiveEvent(calendarEvent)) 
    }

    const startSavingEvent = async ( calendarEvent ) =>{
        
        try {
            if ( calendarEvent.id){
                //update active event
                await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent)
                dispatch( onUploadNewEvent({...calendarEvent, user}))
                return
            }
            //create
                const { data } = await calendarApi.post('/events/create', calendarEvent)
                dispatch( onAddNewEvent({ ...calendarEvent, id: data.event.id, user }))
        } catch (error) {
            console.log(error)
            Swal.fire("Error saving event", error.response.data.msg,'error')
        }
    }
//solo edite algo de aca y funciono
    const startDeletingEvent = () =>{
        dispatch(onDeleteEvent())
    }

    const startLoadingEvents = async() =>{
        try {
            const { data } = await calendarApi.get('/events/get')
            
            const events = convertEventDate( data.event)
            dispatch( onLoadEvents(events) )

        } catch (error) {
            console.log("There was an error loading events")
            console.log(error)
        }
    }
    
    return {
        //properties
        events,
        activeEvent,
        hasEventSelected: !!activeEvent,
        //methods
        setActiveEvent,
        startDeletingEvent,
        startLoadingEvents,
        startSavingEvent,
    }
}