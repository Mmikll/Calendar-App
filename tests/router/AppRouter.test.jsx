import { render, screen } from "@testing-library/react"
import { AppRouter } from "../../src/router/AppRouter"
import { useAuthStore } from "../../src/hooks/useAuthStore"
import { MemoryRouter } from "react-router"
import { CalendarPage } from "../../src/calendar"


jest.mock('../../src/hooks/useAuthStore')
jest.mock('../../src/calendar', () =>({ //I did thios mock so I don't need to mock all the reducers in the calendar page.
    CalendarPage: () => <h1>CalendarPage</h1>
}))

describe('test in <AppRouter/>', () => { 

    const mockCheckAuthToken = jest.fn()
    beforeEach( () => jest.clearAllMocks())
    
    test('should show loading page and have called checkAuthToken', () => { 
        useAuthStore.mockReturnValue({
            status: 'checking',
            checkAuthToken: mockCheckAuthToken,
        })
        render( <AppRouter/>)

        expect(screen.getByText('Loading...')).toBeTruthy()
        expect(mockCheckAuthToken).toHaveBeenCalled()
    })

    test('should login page if user is not authenticated', () => { 
        useAuthStore.mockReturnValue({
            status: 'not-authenticated',
            checkAuthToken: mockCheckAuthToken,
        })
        
        const { container } = render( 
            <MemoryRouter initialEntries={['/auth2/something/some']}>
                <AppRouter/>
            </MemoryRouter>
        )

        expect( screen.getByText('SignIn')).toBeTruthy()
        expect(container).toMatchSnapshot()

    })

    test('should calendar page if user is authenticated', () => { 
        useAuthStore.mockReturnValue({
            status: 'authenticated',
            checkAuthToken: mockCheckAuthToken,
        })
        
        render( 
            <MemoryRouter>
                <AppRouter/>
            </MemoryRouter>
        )

        expect( screen.getByText('CalendarPage')).toBeTruthy()


    })


})