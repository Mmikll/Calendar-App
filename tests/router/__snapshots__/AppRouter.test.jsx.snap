// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <AppRouter/> should calendar page if user is authenticated 1`] = `
<div>
  <div>
    <div
      class="container login-container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 login-form-1 animate__animated animate__bounceInLeft"
        >
          <h3>
            SignIn
          </h3>
          <form>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="loginEmail"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="loginPassword"
                placeholder="Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="d-grid gap-2"
            >
              <input
                class="btnSubmit"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
        <div
          class="col-md-6 login-form-2 animate__animated animate__bounceInRight "
        >
          <h3>
            Register
          </h3>
          <form>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerName"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerEmail"
                placeholder="Email"
                type="Email"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerPassword"
                placeholder="Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerPassword2"
                placeholder="Confirm your Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="btnSubmit"
                type="submit"
                value="Create Account"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test in <AppRouter/> should login page if user is not authenticated 1`] = `
<div>
  <div>
    <div
      class="container login-container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 login-form-1 animate__animated animate__bounceInLeft"
        >
          <h3>
            SignIn
          </h3>
          <form>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="loginEmail"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="loginPassword"
                placeholder="Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="d-grid gap-2"
            >
              <input
                class="btnSubmit"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
        <div
          class="col-md-6 login-form-2 animate__animated animate__bounceInRight "
        >
          <h3>
            Register
          </h3>
          <form>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerName"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerEmail"
                placeholder="Email"
                type="Email"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerPassword"
                placeholder="Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerPassword2"
                placeholder="Confirm your Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="btnSubmit"
                type="submit"
                value="Create Account"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test in <AppRouter/> should login page if user is not authenticated 2`] = `
<div>
  <div>
    <div
      class="container login-container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 login-form-1 animate__animated animate__bounceInLeft"
        >
          <h3>
            SignIn
          </h3>
          <form>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="loginEmail"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="loginPassword"
                placeholder="Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="d-grid gap-2"
            >
              <input
                class="btnSubmit"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
        <div
          class="col-md-6 login-form-2 animate__animated animate__bounceInRight "
        >
          <h3>
            Register
          </h3>
          <form>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerName"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerEmail"
                placeholder="Email"
                type="Email"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerPassword"
                placeholder="Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="form-control"
                name="registerPassword2"
                placeholder="Confirm your Password"
                type="Password"
                value=""
              />
            </div>
            <div
              class="form-group mb-2"
            >
              <input
                class="btnSubmit"
                type="submit"
                value="Create Account"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
